tooling/nargo_cli/tests/execution_success/trait_where_clause/src/main.nr

```diff
@@ -0,0 +1,41 @@
+// TODO(#2568): Currently we only support trait constraints on free functions.
+// There's a bunch of other places where they can pop up:
+//      - trait methods (trait Foo<T> where T: ... { )
+//      - free impl blocks (impl<T> Foo<T> where T...)
+//      - trait impl blocks (impl<T,U> Foo for Bar where T...)
+//      - structs (struct Foo<T> where T: ...)
```

Comment:
It's generally not best practice to put trait constraints directly on structs anyway, so I don't think you need to worry about that. It may not be something we want at all. Trait constraints on `trait` declarations themselves (the first bullet point) also aren't the most useful.

Reply:
Ok, I'll note to focus on trait constraints for the impl blocks first. At a first glance it seems to me like those should be as easy as "add the constraints of the impl block to all generic methods in it", maybe I should've even done it in this PR.